---
title: "macrosyntR"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{macrosyntR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}

library(macrosyntR)

```

We demonstrate the usage of the package using publicly available data.   

* Branchiostoma floridae (Simakov et al. 2020) (data download : https://www.ncbi.nlm.nih.gov/genome/?term=txid7739)   

* Paraescarpia echinospica (Sun et al. 2021)
(data download : https://doi.org/10.6084/m9.figshare.15050478.v1)

# Pre-process and Load the data using load_mbh_df() :

### Theory :

This package doesn't compute the orthologs. I recommend to compute it as mutual best hits.   
It is quite fast and accurate when using diamond blast, In case you have no idea on how to do it,
I made available on github a shell script that would do it for you (see here : [mbhXpress](https://github.com/SamiLhll/mbhXpress))

Drawing the plots require to have the the following values for each pair of genes :
* genomic coordinates on species 1
* genomic coordinates on species 2

As any method to get the orthologs between two sets of protein sequences will likely report the sequences unique identifiers, this package provides a method that use these ids to link the BED formatted genomic coordinates of the two species for each orthologous pair.

let's say I have the following orthologs :   
sp1.gene.x1  sp2.gene.y1   
sp1.gene.X2  sp2.gene.y2   
...   
sp1.gene.xn  sp2.gene.yn   

the following genomic coordinates for species1 :   
chr4    200    600    sp1.gene.x1   
chr8     10    400    sp1.gene.x2   
...   
chr12   900    980    sp1.gene.xn   

and the following genomic coordinates for species2 :
chr4    200    600    sp1.gene.x1   
chr8     10    400    sp1.gene.x2   
...   
chr12   900    980    sp1.gene.xn  

### Example :

The next functions would work if you input a dataframe shaped as the one produced by load_mbh_df().



